pkg/crd/crds_test.go

```diff
@@ -0,0 +1,239 @@
+	if err != nil {
+		t.Fatal("Failed to convert CRD to apiextensionsv1.CustomResourceDefinition")
+	}
+
+	bpCRD := apiextensions.CustomResourceDefinition{}
+	err = apiextensionsv1.Convert_v1_CustomResourceDefinition_To_apiextensions_CustomResourceDefinition(&bpCRDv1, &bpCRD, nil)
+	if err != nil {
+		t.Fatal("Failed to convert CRD to apiextensions.CustomResourceDefinition")
+	}
+
+	// We assume there is only one version in the CRD
+	ver := bpCRD.Spec.Versions[0]
+	crdSchema := ver.Schema
+	if crdSchema == nil {
+		crdSchema = bpCRD.Spec.Validation
+	}
+
+	schemaValidator, _, err := validation.NewSchemaValidator(crdSchema.OpenAPIV3Schema)
+	if err != nil {
+		t.Fatal("Failed to create a validator")
+	}
+
+	testCases := []struct {
+		name     string
+		manifest string
+		wantErr  bool
+	}{
+		{
+			name:     "BindingPolicy with invalid matchLabels value",
+			manifest: bp_bad_matchLabels_value,
+			wantErr:  true,
+		},
+		{
+			name:     "BindingPolicy with valid matchLabels",
+			manifest: bp_good_matchLabels,
+			wantErr:  false,
+		},
+		{
+			name:     "BindingPolicy with invalid matchExpressions key",
+			manifest: bp_bad_matchExpressions_key,
+			wantErr:  true,
+		},
+		{
+			name:     "BindingPolicy with invalid matchExpressions operator",
+			manifest: bp_bad_matchExpressions_op,
+			wantErr:  true,
+		},
+		{
+			name:     "BindingPolicy with invalid matchExpressions value",
+			manifest: bp_bad_matchExpressions_value,
+			wantErr:  true,
+		},
+		{
+			name:     "BindingPolicy with valid matchExpressions",
+			manifest: bp_good_matchExpressions,
+			wantErr:  false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			err := validateCustomResourceYaml(tc.manifest, schemaValidator)
+			if tc.wantErr != (err != nil) {
+				t.Errorf("wantErr: %v, error, %s", tc.wantErr, err)
+			} else if err != nil {
+				t.Logf("Got expected validation error, %s", err)
+			}
+		})
+	}
+}
+
+func validateCustomResourceYaml(data string, vd validation.SchemaValidator) error {
+	cr := &unstructured.Unstructured{}
+	if err := yaml.Unmarshal([]byte(data), cr); err != nil {
+		return err
+	}
+	if err := validation.ValidateCustomResource(nil, cr.Object, vd).ToAggregate(); err != nil {
+		return fmt.Errorf("%v %v is invalid, err: %w", cr.GetKind(), cr.GetName(), err)
+	}
+	return nil
+}
+
+var bp_bad_matchLabels_value = `
+apiVersion: control.kubestellar.io/v1alpha1
+kind: BindingPolicy
+metadata:
+  name: nvidia-gpu-placement
+spec:
+  clusterSelectors:
+  - matchLabels:
+      name: edgeplatform-test-wec1, edgeplatform-test-wec2
+  downsync:
+  - objectSelectors:
+    - matchLabels:
+        app.kubernetes.io/part-of: wec-gpu
+  wantSingletonReportedState: true
+`
+
+var bp_good_matchLabels = `
+apiVersion: control.kubestellar.io/v1alpha1
+kind: BindingPolicy
+metadata:
+  name: nvidia-gpu-placement
+spec:
+  clusterSelectors:
+  - matchLabels:
+      name: edgeplatform-test-wecs
```

Comment:
The "good" cases should be more challenging. Reviewing https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set, I suggest something like the following.

```
  - matchLabels: {"sub.domain.test/0.9-5_Xy": "This-is_1.test"}
```

Reply:
Changed as suggested.